django_app/redbox_app/redbox_core/consumers.py

```diff
@@ -255,10 +265,30 @@ async def handle_activity(self, response: dict):
         self.activities.append(RedboxActivityEvent.model_validate(response))
 
     async def handle_documents(self, response: list[Document]):
-        s3_keys = [doc.metadata["file_name"] for doc in response]
-        files = File.objects.filter(original_file__in=s3_keys)
-
+        sources_by_resource_ref = {}
+        for document in response:
+            ref = document.metadata.get("original_resource_ref")
+            docs_from_source = sources_by_resource_ref.get(ref, [])
+            docs_from_source.append(document)
+            sources_by_resource_ref[ref] = docs_from_source
+        handled_sources = set()
+        files = File.objects.filter(original_file__in=sources_by_resource_ref.keys())
         async for file in files:
             await self.send_to_client("source", {"url": str(file.url), "original_file_name": file.original_file_name})
-        for file in files:
-            self.citations.append((file, [doc for doc in response if doc.metadata["file_name"] == file.unique_name]))
+            self.citations.append(
+                (file, sources_by_resource_ref[file.unique_name])
+            )
+            handled_sources.add(file.unique_name)
+
+        additional_sources = [doc for doc in response if doc.metadata["original_resource_ref"] not in handled_sources]
```

Comment:
tbh i dont get this logic - what is special about the first citation that only it is treated as if it has a file?